presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/StaticHiveCluster.java

```diff
@@ -67,12 +82,27 @@ public StaticHiveCluster(List<URI> metastoreUris, String metastoreUsername, Hive
     public HiveMetastoreClient createMetastoreClient()
     {
         List<HostAndPort> metastores = new ArrayList<>(addresses);
-        Collections.shuffle(metastores.subList(1, metastores.size()));
+        metastores.sort(Comparator.comparingLong(m -> concurrentMap.get(m).getBackoffDelayNanos()));
 
         TException lastException = null;
         for (HostAndPort metastore : metastores) {
             try {
-                HiveMetastoreClient client = clientFactory.create(metastore);
+                HiveMetastoreClient client = new FailureAwareHiveMetaStoreClient(clientFactory.create(metastore), new ResponseHandle()
+                {
+                    @Override
+                    public void success()
+                    {
+                        concurrentMap.get(metastore).success();
+                    }
+
+                    @Override
+                    public void failed(TException e)
+                    {
+                        if (e.getCause() instanceof SocketTimeoutException) {
+                            concurrentMap.get(metastore).fail();
```

Comment:
I don't understand this exp backoff logic. It looks natural at first... Yet, please help me understand in what situations is the implemented logic better than something like "when metastore fails, move its address to the end of the queue" ?
when there is some sustained load (e.g. few queries caught in RetryDriver loop because metastores fail), you ultimately pick up the metastore instance that failed least recently, but you could express this logic in far simpler way

Reply:
yeah. I agree moving to end of queue is a better approach. I don't think of differences off top of my head. let me refine this.

Reply:
If this works out, this would greatly simplify the code: no need for anything time-related

Reply:
I do think of one minor difference. the backoff approach could satisfy "prefer working metastore according to the specified order". The queue approach will never try failed one if current one is fine. However, this feature might not be very useful. your option?

Reply:
My opinion is: the simpler the better, unless we know experience that this simpler is too simple.

Either way, we no longer pick up first Metastore address by default. This needs to be double-checked with FB.

Reply:
ok. sure. I will simplify this then.